### PR TITLE
Add article about unsigned integers

### PR DESCRIPTION
**Pull request by Eisenwave**
Closes server request https://discord.com/channels/331718482485837825/802541516655951892/1304467656530984981

image

### CODE REVIEW

File: articles/unsigned.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
+# Why Should You Avoid `unsigned` for Non-Negative Parameters?
```
[the remark this answers is not in the document]
Comment (Eisenwave) [reply articles/unsigned.md:1]: It was the motivation for writing the article in the first place, but it doesn't justify its existence within the article.

**Comment by Eisenwave** [comment]
image

File: articles/unsigned.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+```cpp
+auto deficit = target_size - size();
+if (deficit > 0) { add_amount(deficit); }
+else             { remove_amount(deficit); }
+```
+**Bug:**: Subtraction wraps around, so `deficit`
+is never negative.
+
+```cpp
+if (i > -1)
```
[the remark this answers is not in the document]
Comment (Eisenwave) [reply articles/unsigned.md]: I've added
```cpp
unsigned i = /* ... */;
```
... which saves one line.

It would be awkward to open a function and then not have a closing brace in the code block or something, so this seems like the most compact approach.

File: articles/unsigned.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
+# Why Should You Avoid Unsigned Integers?
+
+Unsigned integers should generally be avoided because they have
+[modular arithmetic][mod] (wrap around),
+leading to surprising results:
```
Comment (jeremy-rifkin) [inline articles/unsigned.md:5]: It might be good to frame this in terms of surprising behavior with signed-unsigned comparison/conversion instead, that's really where the bug-proneness lies.

File: articles/unsigned.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+```cpp
+my_vector[negative_int] = /* ... */; 
+```
+**Bug:**: `negative_int` is implicitly converted to a huge positive integer,
+but `operator[]` thinks we cannot pass negatives.
```
Comment (jeremy-rifkin) [inline articles/unsigned.md:11]: `my_vec[-1]` would be a bug regardless of the `operator[]`'s parameter, something like this might work a little better to highlight how the unsigned conversion here can be uniquely problematic

````suggestion
```cpp
void foo(unsigned val) { std::print("{}", val); }
int main() { foo(-1); } // prints 4294967295
```
**Bug:** `-1` is implicitly converted to a huge positive integer.
````

(Also fixed a super minor typo)

File: articles/unsigned.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+```cpp
+auto deficit = target_size - size();
+if (deficit > 0) { add_amount(deficit); }
+else             { remove_amount(deficit); }
+```
+**Bug:**: Subtraction wraps around, so `deficit`
```
Comment (jeremy-rifkin) [inline articles/unsigned.md:18]:
```suggestion
**Bug:** Subtraction wraps around, so `deficit`
```

File: articles/unsigned.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+if (i > -1) { /* ... */ }
+```
+**Bug:**: The if statement is never entered because `-1` is
+converted to
+[`UINT_MAX`](https://en.cppreference.com/w/c/types/limits).
+
+<!-- inline -->
+## Recommendation
+
+Use unsigned integers only for bit manipulation or I/O tasks;
```
Comment (jeremy-rifkin) [inline articles/unsigned.md:32]: I think I/O here could be better specified as "serialization." If a beginner is reading this and if they know what "I/O" means at all they are probably thinking iostreams and `std::cin`.

File: articles/unsigned.md [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+prefer signed integers otherwise,
+and avoid mixing signed/unsinged above all else.
+
+Compiler warnings can be used to detect the issues above
+(e.g. [`-Wsign-conversion`](https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wsign-conversion)),
```
Comment (jeremy-rifkin) [inline articles/unsigned.md:37]: Please mention `-Wsign-compare` too

File: articles/unsigned.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+but they detect countless harmless cases.
+
+<!-- inline -->
+## See Also
+
+- [CppCoreGuidelines ES.106][guide] *Don’t try to avoid negative values by using unsigned*
+- [Bjarne Stroustrup - P1428][p1428] *Subscripts and sizes should be signed*
```
Comment (jeremy-rifkin) [inline articles/unsigned.md:44]: I think it'd be worth linking the lightning talk too https://www.youtube.com/watch?v=wvtFGa6XJDU

File: articles/unsigned.md (hunk 3 above)
Comment (Eisenwave) [reply articles/unsigned.md:5]: It's probably true that this is the source of most bugs. Do keep in mind that the article is mainly aimed at people to whom it was intuitive that non-negative function parameters should be `unsigned` instead of `int` and need a little of convincing.

If you frame it entirely as signed/unsigned mixed issues, they can just say
> OK so just use `unsigned` everywhere.

The second example
```cpp
auto deficit = target_size - size();
```
... not only disarms this idea that you're safe if you never use `signed`, but also demonstrates that returning an `unsigned`, non-negative quantity can go wrong.

In short, I feel like `unsigned` numbers are problematic in general and we shouldn't frame it like avoiding mixing signed/unsigned will fix them.

**Comment by Eisenwave** [comment]
image

**Review by 64 — APPROVED** [review]

File: articles/unsigned.md (hunk 3 above)
Comment (jeremy-rifkin) [reply articles/unsigned.md:5]: I'm fine with merging as is. The main reason I bring this up is that having modular arithmetic behavior isn't all that surprising and seen by many as desirable. Usually people are surprised to learn that signed overflow isn't defined in C++. One wording that could be used here is `Unsigned integers should be avoided due to unintentional wraparound and surprising behavior with signed-unsigned conversion and comparison.` Thoughts?

Comment (Eisenwave) [reply articles/unsigned.md]: Great idea!

**Review by jeremy-rifkin — APPROVED** [review]
LGTM thanks
